File: test/Filters/test-IIRFilter.py
from scipy.signal import lfilter
import numpy as np

type = 'int'
b = np.array((1, 10, 2, -3, -1), dtype=np.float64)
a = np.array((-1, 2, -3), dtype=np.float64)

signal = np.array((100, 10, 102, 23, 51, 1, -10, -53, 100, -100, 100, -10, 10,
                   11, 20, 30, 123, 12, 90, 10),
                  dtype=np.float64)
output = lfilter(b, a, signal)
print(f'IIRFilter<{len(b)}, {len(a)}, {type}> filter = {{{{')
print(' ', ', '.join(map(lambda x: str(x), np.round(b).astype(np.int))))
print('},')
print('{')
print(' ', ', '.join(map(lambda x: str(x), np.round(a).astype(np.int))))
print('}};')
print(f'array<{type}, {len(signal)}> signal = {{')
print(' ', ', '.join(map(lambda x: str(x), np.round(signal).astype(np.int))))
print('};')
print(f'array<{type}, {len(signal)}> expected = {{')
print(' ', ', '.join(map(lambda x: str(x), np.round(output).astype(np.int))))
print('};')
print(
    f'for_each(signal.begin(), signal.end(), [&]({type} &s) {{ s = filter(s); }});'
)
print('EXPECT_EQ(signal, expected);')

File: test/Filters/test-IIRFilter.cpp
#include <gtest/gtest.h>

#include <Filters/IIRFilter.hpp>

TEST(IIRFilter, IIRFilterRandomInt) {
    using namespace std;
    IIRFilter<5, 3, int> filter = {{1, 10, 2, -3, -1}, {-1, 2, -3}};
    array<int, 20> signal = {100, 10,  102, 23, 51, 1,  -10, -53, 100, -100,
                             100, -10, 10,  11, 20, 30, 123, 12,  90,  10};
    array<int, 20> expected = {-100,     -1210,    -2522,   -2177,    2857,
                               12004,    15506,    -4673,   -55360,   -97524,
                               -28437,   235155,   555311,  405266,   -855481,
                               -2926992, -3287961, 2203823, 14271183, 21930362};
    for_each(signal.begin(), signal.end(), [&](int &s) { s = filter(s); });
    EXPECT_EQ(signal, expected);
}

TEST(IIRFilter, IIRFilterRandomDouble) {
    using namespace std;
    IIRFilter<5, 3, double> filter = {{1, 10, 2, -3, -1}, {-1, 2, -3}};
    array<double, 20> signal = {
        100, 10,  102, 23, 51, 1,  -10, -53, 100, -100,
        100, -10, 10,  11, 20, 30, 123, 12,  90,  10,
    };
    array<double, 20> expected = {
        -100,    -1210,    -2522,    -2177,   2857,     12004,    15506,
        -4673,   -55360,   -97524,   -28437,  235155,   555311,   405266,
        -855481, -2926992, -3287961, 2203823, 14271183, 21930362,
    };
    for_each(signal.begin(), signal.end(), [&](double &s) { s = filter(s); });
    EXPECT_EQ(signal, expected);
}